### ISSUE TITLE
Question: How to mock a method that passes a parameter by reference?

### ISSUE THREAD

**Issue by rpottsoh** [VSoftTechnologies/Delphi-Mocks#88]
I have a method (function) that has two arguments and returns an integer.  The first argument is passed by value and the second by reference (var).  I would like to mock what this method should pass back on on the second argument.  First, the closest thing I can see to accomplish this is to use WillExecute.  However this only passes my method's arguments in as const.  I am setting the result of my TExecuteFunc equal to args[0].  I would like to set args[1] equal to some value.

How to mock, or stub, a method that has one or more arguments that are passed by reference?

`    MockDap.Setup.WillExecute(function (const args : TArray<TValue>; const ReturnType : TRttiType) : TValue
                              begin
                                result := args[0];
                                args[1] := Expected;
                              end).When.GetDAPData(TheLength, Actual);
`

Thanks

**Comment by staticcat** [comment]
Currently there isn't a way to achieve var or out parameter return values with Delphi.Mocks. I have had plans on how to achieve this, just haven't had the time to attempt its implementation. 

The DoInvoke of the proxy classes needs to handle updating var variables and out variables. Even though the args array is const itself, from memory the values of the variables at each index could be altered to the desired values. These would then be returned to the caller, only when the value was either listed as var, or out. If there is information on the method as to the parameter type we could even protect against this.

**Comment by rpottsoh** [comment]
Thanks for the reply.  Doesn't sound like any small undertaking to achieve this capability.  My best bet would probably be to eliminate my argument passed by reference and have my method just return the data instead; pass the old result value of the method through a separate property/method. 


Sent from my android device.

-----Original Message-----

**Comment by staticcat** [comment]
Yes, and it is one of the draw backs for using Delphi.Mocks. Even in C# you have to do something like the following for Moq. 

int outVariable = 2;
myMock.Setup(x => x.FuncWithOutVar(out outVariable)).Returns(true);

The outVariable always gets a hint saying it could be removed as the value is not used. I suspect we would get the same thing in delphi as the compiler has no idea we are intercepting the method call.
